### DISCUSSION TITLE
Road towards new alpha

### DISCUSSION THREAD

**Discussion by lushoBarlett** [SebastianMestre/Jasper#208] [Language Feature]
I'm curious as to what we should be doing now. We could do what we did some time ago and make a graph of things we want to do and plot their dependencies and such.

**Comment by SebastianMestre** [comment]
I think the next release should focus on polish and robustness. Here is a small list of things that need addressing

 - think more clearly about the type system, lay out a formal-ish description of it, and start touching up the implementation to match.
 - pipe compilation errors through the compiler, up to the top level, so we can print pretty error reports.
 - #183

**Comment by lushoBarlett** [comment]
Maybe we could also take the type hints of function arguments into account. I was writing some code the other day and the interpreter exploded when the match expression could not figure out what union type it was. I put the type in the argument of the function instead of the match.

**Reply by SebastianMestre** [reply]
Yeah, that's just messy.

**Comment by SebastianMestre** [comment]
Also, as I code, I tend to pepper `// TODO:` comments all around when I'm not sure about something. I try to look into them every once in a while, but I think I we could look at those for places with possible robustness issues or latent bugs.

**Reply by lushoBarlett** [reply]
This is very reasonable if we are going for robustness and general cleanup.

**Comment by lushoBarlett** [comment]
Some other ideas I have, although less important for your approach to the new alpha, are checking out how to improve the declaration and naming and use of native functions, maybe even add input.

**Reply by SebastianMestre** [reply]
I especially agree with the input part.

**Comment by SebastianMestre** [comment]
Another thing I think would be nice is to stop embedding the language integration tests in a .cpp file. Instead, we should have a bunch of jasper files, along with their expected output in a subdirectory, and a test runner script that runs every test and checks for matching output.

This way, we don't need to re-compile a .cpp file everytime we add or modify a test.

(EDIT) of course, not all of our tests fit this framework, so they'd still be embedded in C++.

**Comment by lushoBarlett** [comment]
Oh, we must not forget about #191, maybe we can take over that one and finish it ourselves?

**Comment by SebastianMestre** [comment]
Another thing, we should stop implicitly adding poly parens to identifiers in type expressions

**Reply by SebastianMestre** [reply]
Or is that done already?

**Reply by lushoBarlett** [reply]
Good question, I think not actually.

**Comment by lushoBarlett** [comment]
I remembered that we can't actually parse negative numbers. So maybe that one is an easy addition.

**Comment by SebastianMestre** [comment]
Here is a compiled list. I've separated it into things I think should be in the next alpha, and things that need to be done eventually (which could very well be on the next alpha) (EDIT: moved to #237).

### Next alpha
- [x] #183
- [x] Do function argument type hints
- [x] Propagate errors through the compiler phases, and print them
- [x] Standard library (I/O in particular)
- [x] Don't implicitly add poly parens when parsing type expressions
- [x] Parse negative numbers (or prefix minus operator?)
- [x] Array element assignment
- [ ] Look around `TODO` comments

**Comment by SebastianMestre** [comment]
From tomorrow, imma start working on error propagation and reporting.

I have two different ideas for how that could work. I'm in a bit of a hurry rn but the high level is as follows:

 1. Return an error value, propagate it and bubble it up to higher stack frames.
 1. Store an error value in the typechecker, and return a status code (or store it in the typechecker).
 1. Throw exceptions and leave it to the caller to catch them.

I think number 2 would be most performant, but I don't really know. I wanna see which one will give us cleaner code and be less error prone.

The most challenging thing here is probably extracting error data from the unification cores. I have two ideas for this, too

 1. store a `TypedAST*` and some metadata on every unification node, to be able to tell where it comes from, and return an error code along with the appropriate metadata when there is an eror. This couples the unification code with the rest of the compiler, which I don't love.
 1. Add the notion of "Error nodes" which indicate there was an error during unification or creation of the node. This type of node will store information about the error, and it will infect any node it interacts with, turning it into another error node (or into a var that points at one).
 
I'm not clear on the details for either idea, but it's a good place to start from. (Maybe the two can be combined, too)

Any opinions? @lushoBarlett

**Comment by SebastianMestre** [comment]
Oh, I forgot to mention: it would be great to make static type checking optional in the future.

(EDIT)

After thinking about it, it seems to me like it's actually pretty easy to do: we currently rely on type inference to carry type information to the point where we use constructors, where we grab that information from within the typechecker bowels.

As it turns out, this is not necessary: if we had access to the `StructExpression/UnionExpression` nodes we could easily get that information from them, instead.

In the current implementation, we replace those nodes with `xxxHandles` in `ct_eval`. This turns out to be a non-issue, as the handles store a pointer to the olds nodes anyways, meaning we can still do thing we want, given a Handle node.

**Comment by SebastianMestre** [comment]
I think we could also fit array access assignment into the next alpha, no?

It'd be an interpreter-only change, as the type system already allows this.

```rust
x := seq {
  arr := array { 0; 0; 0; 0; };
  arr[3] = 1; // this would be new
  return arr;
};
```

**Reply by SebastianMestre** [reply]
Done in #224

**Comment by SebastianMestre** [comment]
I'm working on improving error messages over on #229. I'll check that off the list, as I believe that's good enough for now.

Going forward, I would like a way to pass a location in jasper code to the logging library to show where the error comes from, though.

(edit) also, we need to integrate `ErrorReport` with the recently created logging library

**Comment by SebastianMestre** [comment]
alpha 3 is pretty much feature-complete I'd say.
